internal/provider/trigger_resource.go

```diff
@@ -385,10 +489,107 @@ func (r *triggerResource) ImportState(ctx context.Context, req resource.ImportSt
 	resp.Diagnostics.Append(resp.State.Set(ctx, &models.TriggerResourceModel{
 		ID:         types.StringValue(id),
 		Dataset:    types.StringValue(dataset),
+		QueryID:    types.StringNull(),
+		QueryJson:  types.StringUnknown(), // favor QueryJSON on import
 		Recipients: types.SetUnknown(types.ObjectType{AttrTypes: models.NotificationRecipientAttrType}),
 	})...)
 }
 
+func (r *triggerResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	var data models.TriggerResourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// exit early if we don't have QueryJSON
+	if data.QueryJson.IsNull() || data.QueryJson.IsUnknown() {
+		return
+	}
```

Comment:
Does returning early here mean it successfully validated or that it failed?

Reply:
Neither: we can't validate anything that isn't there

Reply:
Ok, maybe this is a mental model fail on my part -- I was thinking about the line above

```go
if resp.Diagnostics.HasError() {
		return
	}
```
which to me read a bit like "if there's an existing error the config of the trigger resource fails validation". But maybe it means something closer to "if there's an existing error [or we don't have queryJSON]", don't run further validation"? If that's right, that makes sense 👍

Reply:
That's a different error case: it means we were unable to get the configuration so we fail out right away.

These functions don't have returned results (no boolean or error), but instead add to a `*resource.ValidateConfigResponse` -- often with the intention of adding more than one error or warning -- but there are also cases where you know it's not safe to continue.